### ISSUE TITLE
Compiler crash with generic ffi types

### ISSUE THREAD

**Issue by simolus3** [dart-lang/sdk#39444]
Running the following Dart file, which is accepted by the analyzer, crashes the compiler. I'm not entirely sure if the file is valid, but I think the error message should at least provide a location.
```dart
import 'dart:collection';
import 'dart:ffi';

import 'package:collection/collection.dart';
import 'package:ffi/ffi.dart';

/// A [List] backed by natively-managed memory.
class NativeList<T extends Struct>
    with ListMixin<T>, NonGrowableListMixin<T>
    implements List<T> {
  @override
  final int length;
  final int _elementSize;
  final Pointer<T> _address;

  NativeList._(this.length, this._address) : _elementSize = sizeOf<T>();

  NativeList.fromPointerAndLength(Pointer<T> ptr, int length)
      : this._(length, ptr);

  factory NativeList.allocate(int length) {
    return NativeList<T>.fromPointerAndLength(
        allocate<T>(count: length), length);
  }

  factory NativeList.allocateFrom(List<T> list) {
    final native = NativeList<T>.allocate(list.length);
    native.setAll(0, list);
    return native;
  }

  Pointer<T> get pointer => _address;

  @override
  T operator [](int index) {
    return _address.elementAt(index).ref;
  }

  @override
  void operator []=(int index, T value) {
    final targetBytes = _address.elementAt(index).cast<Uint8>();
    final structBytes = value.addressOf.cast<Uint8>();

    for (var j = 0; j < _elementSize; j++) {
      targetBytes[j] = structBytes[j];
    }
  }
}

class TestStruct extends Struct {
  @Int32()
  int foo;
  @Int32()
  int bar;
}

void main() {
  final firstPtr = allocate<TestStruct>();
  final secondPtr = allocate<TestStruct>();

  final first = firstPtr.ref;
  final second = secondPtr.ref;

  first
    ..foo = 3
    ..bar = 4;
  second
    ..foo = 6
    ..bar = 7;

  final list = NativeList<TestStruct>.allocateFrom([first, second]);
}
```

Output of running `dart <file.dart>`:
```
Crash when compiling null,
at character offset null:
type 'DynamicType' is not a subtype of type 'InterfaceType' in type cast
#0      _FfiUseSiteTransformer.visitMethodInvocation (package:vm/transformations/ffi_use_sites.dart:321:47)
#1      MethodInvocation.accept (package:kernel/ast.dart:3254:44)
#2      transformList (package:kernel/ast.dart:6912:27)
#3      Arguments.transformChildren (package:kernel/ast.dart:3150:5)
#4      Transformer.defaultTreeNode (package:kernel/visitor.dart:640:10)
#5      TreeVisitor.visitArguments (package:kernel/visitor.dart:261:39)
#6      Arguments.accept (package:kernel/ast.dart:3140:38)
#7      StaticInvocation.transformChildren (package:kernel/ast.dart:3373:29)
#8      Transformer.defaultTreeNode (package:kernel/visitor.dart:640:10)
#9      TreeVisitor.defaultExpression (package:kernel/visitor.dart:144:43)
#10     TreeVisitor.visitStaticInvocation (package:kernel/visitor.dart:162:53)
#11     _FfiUseSiteTransformer.visitStaticInvocation (package:vm/transformations/ffi_use_sites.dart:123:11)
#12     StaticInvocation.accept (package:kernel/ast.dart:3362:44)
#13     ReturnStatement.transformChildren (package:kernel/ast.dart:4893:31)
#14     Transformer.defaultTreeNode (package:kernel/visitor.dart:640:10)
#15     TreeVisitor.defaultStatement (package:kernel/visitor.dart:203:41)
#16     TreeVisitor.visitReturnStatement (package:kernel/visitor.dart:220:51)
#17     ReturnStatement.accept (package:kernel/ast.dart:4883:43)
#18     transformList (package:kernel/ast.dart:6912:27)
#19     Block.transformChildren (package:kernel/ast.dart:4460:5)
#20     Transformer.defaultTreeNode (package:kernel/visitor.dart:640:10)
#21     TreeVisitor.defaultStatement (package:kernel/visitor.dart:203:41)
#22     TreeVisitor.visitBlock (package:kernel/visitor.dart:206:31)
#23     Block.accept (package:kernel/ast.dart:4452:43)
#24     FunctionNode.transformChildren (package:kernel/ast.dart:2516:19)
#25     Transformer.defaultTreeNode (package:kernel/visitor.dart:640:10)
#26     TreeVisitor.visitFunctionNode (package:kernel/visitor.dart:260:45)
#27     FunctionNode.accept (package:kernel/ast.dart:2500:38)
#28     Procedure.transformChildren (package:kernel/ast.dart:2155:27)
#29     Transformer.defaultTreeNode (package:kernel/visitor.dart:640:10)
#30     TreeVisitor.defaultMember (package:kernel/visitor.dart:230:35)
#31     TreeVisitor.visitProcedure (package:kernel/visitor.dart:232:39)
#32     Procedure.accept (package:kernel/ast.dart:2142:40)
#33     transformList (package:kernel/ast.dart:6912:27)
#34     Class.transformChildren (package:kernel/ast.dart:1266:5)
#35     Transformer.defaultTreeNode (package:kernel/visitor.dart:640:10)
#36     TreeVisitor.visitClass (package:kernel/visitor.dart:239:31)
#37     _FfiUseSiteTransformer.visitClass (package:vm/transformations/ffi_use_sites.dart:85:20)
#38     Class.accept (package:kernel/ast.dart:1208:38)
#39     transformList (package:kernel/ast.dart:6912:27)
#40     Library.transformChildren (package:kernel/ast.dart:539:5)
#41     Transformer.defaultTreeNode (package:kernel/visitor.dart:640:10)
#42     TreeVisitor.visitLibrary (package:kernel/visitor.dart:254:35)
#43     _FfiUseSiteTransformer.visitLibrary (package:vm/transformations/ffi_use_sites.dart:77:18)
#44     List.forEach (dart:core-patch/growable_array.dart:283:8)
#45     transformLibraries (package:vm/transformations/ffi_use_sites.dart:49:13)
#46     VmTarget.performModularTransformationsOnLibraries (package:vm/target/vm.dart:142:5)
#47     KernelTarget.runBuildTransformations (package:front_end/src/fasta/kernel/kernel_target.dart:807:19)
#48     KernelTarget.buildComponent.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:308:7)
<asynchronous suspension>
#49     withCrashReporting (package:front_end/src/fasta/crash.dart:122:24)
<asynchronous suspension>
#50     KernelTarget.buildComponent (package:front_end/src/fasta/kernel/kernel_target.dart:298:12)
<asynchronous suspension>
#51     generateKernelInternal.<anonymous closure> (package:front_end/src/kernel_generator_impl.dart:163:38)
<asynchronous suspension>
#52     withCrashReporting (package:front_end/src/fasta/crash.dart:122:24)
<asynchronous suspension>
#53     generateKernelInternal (package:front_end/src/kernel_generator_impl.dart:69:10)
<asynchronous suspension>
#54     generateKernel.<anonymous closure> (package:front_end/src/kernel_generator_impl.dart:49:18)
<asynchronous suspension>
#55     CompilerContext.runWithOptions.<anonymous closure> (package:front_end/src/fasta/compiler_context.dart:136:20)
<asynchronous suspension>
#56     CompilerContext.runInContext.<anonymous closure>.<anonymous closure> (package:front_end/src/fasta/compiler_context.dart:124:46)
#57     new Future.sync (dart:async/future.dart:224:31)
#58     CompilerContext.runInContext.<anonymous closure> (package:front_end/src/fasta/compiler_context.dart:124:19)
#59     _rootRun (dart:async/zone.dart:1124:13)
#60     _CustomZone.run (dart:async/zone.dart:1021:19)
#61     _runZoned (dart:async/zone.dart:1516:10)
#62     runZoned (dart:async/zone.dart:1463:12)
#63     CompilerContext.runInContext (package:front_end/src/fasta/compiler_context.dart:123:12)
#64     CompilerContext.runWithOptions (package:front_end/src/fasta/compiler_context.dart:134:10)
#65     generateKernel (package:front_end/src/kernel_generator_impl.dart:48:32)
<asynchronous suspension>
#66     kernelForModule (package:front_end/src/api_prototype/kernel_generator.dart:98:17)
<asynchronous suspension>
#67     SingleShotCompilerWrapper.compileInternal (file:///b/s/w/ir/cache/builder/sdk/pkg/vm/bin/kernel_service.dart:338:17)
<asynchronous suspension>
#68     Compiler.compile.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/vm/bin/kernel_service.dart:162:45)
<asynchronous suspension>
#69     new Future.<anonymous closure> (dart:async/future.dart:176:37)
#70     _rootRun (dart:async/zone.dart:1120:38)
#71     _CustomZone.run (dart:async/zone.dart:1021:19)
#72     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
#73     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
#74     _rootRun (dart:async/zone.dart:1124:13)
#75     _CustomZone.run (dart:async/zone.dart:1021:19)
#76     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:947:23)
#77     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:21:15)
#78     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:382:19)
#79     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
#80     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
```

Output of `dart --version`: `Dart VM version: 2.6.1 (Mon Nov 11 13:12:24 2019 +0100) on "linux_x64"`

**Comment by johnniwinther** [comment]
The problem might have been fixed; I cannot reproduce on `2.7.0-dev.2.1`.

**Comment by simolus3** [comment]
I can't reproduce this on 2.7 either, looks like it was fixed indeed. Thanks for taking a look!
